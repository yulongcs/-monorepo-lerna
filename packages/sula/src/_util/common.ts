import assignWith from 'lodash/assignWith';
import isUndefined from 'lodash/isUndefined';
import isArray from 'lodash/isArray';

function withDefined(dest: any, src: any) {
  return isUndefined(src) ? dest : src;
}

export function assignWithDefined(...args: Record<string, any>[]) {
  return assignWith(...args, withDefined);
}

export function toArray<T>(value?: T | T[] | null): T[] {
  if (value === undefined || value === null) {
    return [];
  }

  // @ts-ignore
  return isArray(value) ? value : [value];
}


