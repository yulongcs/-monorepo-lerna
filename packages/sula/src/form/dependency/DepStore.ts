import { NameListMap } from '../../_util/NameListMap';
import { FieldNameList, FieldNamePath } from '../../types/form';
import {
  TransedDependency,
  TransedDependencies,
  Dependencies,
  DependencyType,
  Dependency,
} from '../../types/dependency';
import { toArray } from '../../_util/common';
import assign from 'lodash/assign';
import isUndefined from 'lodash/isUndefined';
import isObject from 'lodash/isObject';
import isEqual from 'lodash/isEqual';
import { triggerPlugin, triggerActionPlugin } from '../../rope/triggerPlugin';
import { HOOK_MARK } from '../FieldGroupContext';
import { getStoreValue } from '../utils/valueUtil';

export default class DepStore {
  public getName: (name: FieldNamePath) => FieldNameList = null!;

  // 主动触发配置
  // TODO 如果删除了，要动态更新
  public depsByFieldNameList: NameListMap<FieldNameList, TransedDependencies> = new NameListMap();

  public parse(
    fieldConfig: any,
    deps: Dependencies,
    getName: (name: FieldNamePath) => FieldNameList,
  ) {
    const fieldNameList: FieldNameList = getName(fieldConfig.fieldKey || fieldConfig.name);
    Object.keys(deps).forEach((type) => {
      const dependency: Dependency = deps[type as DependencyType];
      const { cases, ...globalDep } = dependency;

      if (cases) {
        cases.forEach((kase, index) => {
          const finalDep: Dependency = assign({}, globalDep, kase);
          this.parseCase(fieldNameList, type, finalDep, fieldConfig, getName, index);
        });
      } else {
        this.parseCase(fieldNameList, type, globalDep, fieldConfig, getName);
      }
    });
  }

  private parseCase(
    fieldNameList: FieldNameList,
    type: DependencyType,
    dependency: Dependency,
    fieldConfig: any,
    getName: (name: FieldNamePath) => FieldNameList,
    caseIndex?: number,
  ) {
    const {
      relates,
      inputs,
      output,
      defaultOutput,
      ignores,
      type: depPlugin,
      autoResetValue,
    } = dependency;

    const relatedFieldNameLists: FieldNameList[] = relates.map((relate) =>
      getName ? getName(relate) : toArray(relate),
    );

    // 被动配置变主动
    relatedFieldNameLists.forEach((relatedFieldNameList) => {
      if (!this.depsByFieldNameList.get(relatedFieldNameList)) {
        this.depsByFieldNameList.set(relatedFieldNameList, Object.create(null));
      }

      // 变为主动存储
      const transedDependency: TransedDependency = {
        name: fieldNameList,
        type: depPlugin,
        relates: relatedFieldNameLists,
        ignores: normalizeIgnores(ignores, relatedFieldNameLists.length),
        inputs,
        output,
        defaultOutput,
        autoResetValue: autoResetValue !== false,
        ...(type === 'source' ? { remoteSource: fieldConfig.remoteSource } : {}),
      };

      if (this.depsByFieldNameList.get(relatedFieldNameList)[type]) {
        if (!isUndefined(caseIndex)) {
          if (caseIndex === 0) {
            this.depsByFieldNameList.get(relatedFieldNameList)[type].push([transedDependency]);
          } else {
            const depsOfType = this.depsByFieldNameList.get(relatedFieldNameList)[type];
            const lastDepOfType = depsOfType[depsOfType.length - 1];
            lastDepOfType.push(transedDependency);
          }
        } else {
          this.depsByFieldNameList.get(relatedFieldNameList)[type].push([transedDependency]);
        }
      } else {
        this.depsByFieldNameList.get(relatedFieldNameList)[type] = [[transedDependency]];
      }
    });
  }

  private getDepForm = (ctx) => {
    const { getFieldValueByFieldKey, setFieldValueByFieldKey } = ctx.form.getInternalHooks(
      HOOK_MARK,
    );
    return {
      getFieldValue: getFieldValueByFieldKey,
      setFieldValue: setFieldValueByFieldKey,
    };
  };

  private getDepCtx = (ctx) => {
    return assign({}, ctx, { form: assign({}, ctx.form, this.getDepForm(ctx)) });
  };

  public triggerDependency(originCtx, depsOfType: TransedDependencies, cascadePayload) {
    const ctx = this.getDepCtx(originCtx);
    const { form } = ctx;
    const { getPrevFieldNameByFieldKey, isNeverUpdate } = form.getInternalHooks(HOOK_MARK);

    Object.keys(depsOfType).forEach((type) => {
      const allDeps: TransedDependency[][] = depsOfType[type as DependencyType];

      allDeps.forEach((deps) => {
        for (let i = 0, len = deps.length; i < len; i += 1) {
          const isLastCase = i === deps.length - 1;
          const dep = deps[i];
          const {
            name: affectedFieldNameList,
            relates,
            inputs,
            ignores,
            output,
            defaultOutput,
            type: depPlugin,
            autoResetValue,
          } = dep;

          if (type === 'value') {
            // 如果受影响的fieldName在主动设置store里则不出值关联
            const isWilling = isWillingSetValue(affectedFieldNameList, cascadePayload);
            if (isWilling) {
              return true;
            }
          }

          const values = [];
          let canDep: boolean = false;

          for (let i = 0, len = relates.length; i < len; i += 1) {
            // 这是 fieldKeyList
            const relatedFieldNameList = relates[i];
            const relatedFieldValue = form.getFieldValue(relatedFieldNameList);
            const prevRelatedFieldValue = getStoreValue(
              cascadePayload.cascadePrevStore,
              getPrevFieldNameByFieldKey(relatedFieldNameList),
            );
            // 1. 第一次创建的
            // 2. 前后两次值不一样
            if (
              (isNeverUpdate && isNeverUpdate(relatedFieldNameList)) ||
              prevRelatedFieldValue !== relatedFieldValue
            ) {
              canDep = true;
            }
            values.push(relatedFieldValue);
          }

          /**
           * 不能直接通过值没变就直接返回，例如拷贝（或者有初始值）动态添加的场景
           */
          if (canDep === false) {
            return;
          }

          if (depPlugin) {
            if (type === 'source' && autoResetValue) {
              clearValueForSourceDependency(affectedFieldNameList, ctx, cascadePayload);
            }

            const depPluginCtx = assign({}, ctx, {
              name: affectedFieldNameList,
              values,
              relates,
            });

            const isMatched = triggerPlugin('dependency', depPluginCtx, dep);

            // 如果显示返回false则表示匹配失败
            if (isMatched !== false) {
              return;
            }
          }

          let fn;

          if (type === 'source') {
            fn = autoResetValue
              ? (_name: FieldNameList, _source: any) => {
                  clearValueForSourceDependency(_name, ctx, cascadePayload);
                  form.setFieldSource(_name, _source);
                }
              : form.setFieldSource;
          } else if (type === 'disabled') {
            fn = (_name: FieldNameList, _disabled: boolean) => {
              const prevDisabled = form.getFieldDisabled(_name);
              if (prevDisabled !== _disabled) {
                // 减少不必要的渲染
                form.setFieldDisabled(_name, _disabled);
              }
            };
          } else if (type === 'visible') {
            fn = (_name: FieldNameList, _visible: boolean) => {
              const prevVisible = form.getFieldVisible(_name);
              if (prevVisible !== _visible) {
                // 减少不必要的渲染
                form.setFieldVisible(_name, _visible);
              }
            };
          } else {
            fn = form.setFieldValue;
          }

          const isIgnored = singleMatch(ignores, values);
          if (isIgnored) {
            if (isLastCase) {
              fn(affectedFieldNameList, defaultOutput);
            }

            // 如果忽略但非最后一个case则跳过
            continue;
          }

          /**
           * 远程数据源关联只执行一个case
           */
          if (type === 'source' && !inputs && dep.remoteSource) {
            return this.cascadeSource(ctx, dep, cascadePayload);
          }

          const isAllMatched = inputs && allMatch(inputs, values);
          // 第一个匹配的终止
          if (isAllMatched) {
            fn(affectedFieldNameList, output);
            return;
          }

          // 均为匹配到，且是最后一个case，使用defaultOutput
          if (isLastCase) {
            fn(affectedFieldNameList, defaultOutput);
          }
        }
      });
    });
  }

  private cascadeSource(ctx, dep: TransedDependency, cascadePayload) {
    const { form } = ctx;

    const { name, relates, remoteSource, autoResetValue } = dep;

    if (dep.inputs || dep.type) return; // 非远程数据源场景

    const fetchDepInfo = (relates as FieldNameList[]).reduce((memo, relatedFieldNameList) => {
      const value = form.getFieldValue(relatedFieldNameList);
      // memo[relatedFieldNameList.join('.')] = value;
      memo[relatedFieldNameList[relatedFieldNameList.length - 1]] = value;
      return memo;
    }, {} as Record<string, any>);

    autoResetValue && clearValueForSourceDependency(name, ctx, cascadePayload);
    return triggerActionPlugin(
      ctx,
      assign({}, remoteSource, { params: assign({}, remoteSource.params, fetchDepInfo) }),
      ).then((source: any) => {
      form.setFieldSource(name, source);
    });
  }
}

function clearValueForSourceDependency(fieldNameList: FieldNameList, ctx, cascadePayload) {
  const { form } = ctx;
  const { getFieldNameByFieldKey } = form.getInternalHooks(HOOK_MARK);
  /** fieldKey 转换成 fieldName，来判断value在不在cascadePayload里 */
  const fieldNamePathForValue = getFieldNameByFieldKey(fieldNameList);
  const isWilling = isWillingSetValue(fieldNamePathForValue, cascadePayload);
  if (!isWilling) {
    form.setFieldValue(fieldNameList, undefined);
  }
}

function isWillingSetValue(fieldNameList, cascadePayload) {
  const { cascadeTrigger, cascadeStore } = cascadePayload;

  const isWilling =
    (cascadeTrigger === 'setFieldsValue' || cascadeTrigger === 'setFields') &&
    hasOwnPropWithNameList(fieldNameList, cascadeStore);

  return isWilling;
}

/**
 * 不清空 - (ctx.cascadeTrigger === setFieldsValue || setFields) && values.hasOwnProperty(x);
 * 否则清空
 */
export function hasOwnPropWithNameList(
  fieldNameList: FieldNameList,
  payload: Record<string, Record<string, any>>,
) {
  const len = fieldNameList.length;
  let recurValue = payload;
  for (let i = 0; i < len; i += 1) {
    const namePath = fieldNameList[i];
    if (isObject(recurValue) && recurValue.hasOwnProperty(namePath)) {
      recurValue = recurValue[namePath];
    } else {
      // 提前退出
      return false;
    }
  }
  return true;
}

// ==================== 匹配inputs / ignores ====================
/**
 * @inputs: [[1,2], [2,3]]
 * @values: [1,4]
 * @result: true
 */
export function singleMatch(inputs, values) {
  return inputs.some((input, index) => equal(input, values[index]));
}

/**
 * @rules: [[1,2], [2,3]]
 * @values: [1,4]
 * @result: false
 */
export function allMatch(inputs, values) {
  return !inputs.some((input, index) => !equal(input, values[index]));
}

/**
 *
 * @inputs [1,2]
 * @value 1
 * @return true
 */
export function equal(inputs, value) {
  if (inputs === '*') {
    return true;
  }
  for (let i = 0, len = inputs.length; i < len; i += 1) {
    const input = inputs[i];
    if (isEqual(input, value)) {
      return true;
    }
  }

  return false;
}

function normalizeIgnores(ignores: any[][] | undefined, num: number) {
  if (!ignores) {
    return fill(num, [undefined]);
  }

  return ignores;
}

function fill(num: number, value: any) {
  const result = [];
  for (let i = 0; i < num; i += 1) {
    result[i] = value;
  }

  return result;
}
