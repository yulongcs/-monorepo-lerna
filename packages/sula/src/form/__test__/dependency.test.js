import DepStore, {
  hasOwnPropWithNameList,
  singleMatch,
  allMatch,
  equal,
} from '../dependency/DepStore';
import FormDependency from '../dependency';
import '../../__tests__/common';

const getName = (name) => [name];

describe('depstore', () => {
  describe('store', () => {
    let store;

    beforeEach(() => {
      store = new DepStore();
    });

    it('basic', () => {
      const fields = [
        {
          name: 'input',
          label: 'input',
          field: 'input',
        },
        {
          name: 'input2',
          label: 'input2',
          field: 'input',
          remoteSource: {
            url: '/source.json',
            method: 'post',
          },
          dependency: {
            value: {
              relates: ['input'],
              inputs: [['aaa']],
              output: 'one',
              defaultOutput: '',
            },
            visible: {
              relates: ['input'],
              inputs: [['hidden']],
              output: false,
              defaultOutput: true,
            },
            disabled: {
              relates: ['input'],
              inputs: [['disabled']],
              output: true,
              defaultOutput: false,
            },
            source: {
              relates: ['input'],
              inputs: [['source']],
              output: [],
              defaultOutput: [],
            },
          },
        },
      ];

      store.parse(fields[1], fields[1].dependency, getName);
      expect(store.depsByFieldNameList.getNameLists()).toEqual([['input']]);
      const inputType = store.depsByFieldNameList.get(['input']);
      expect(inputType).toMatchSnapshot();

      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'aaa'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'aaa'),
        getFieldVisible: jest.fn(),
        getFieldDisabled: jest.fn(),
        getInternalHooks: () => {
          return depForm;
        },
      };
      store.triggerDependency({ form }, inputType, {});
      expect(form.setFieldValue).not.toHaveBeenCalled();
      expect(depForm.setFieldValueByFieldKey).toHaveBeenCalledWith(['input2'], 'one');
      expect(form.setFieldDisabled).toHaveBeenCalledWith(['input2'], false);
      expect(form.setFieldVisible).toHaveBeenCalledWith(['input2'], true);
      expect(form.setFieldSource).toHaveBeenCalledWith(['input2'], []);
    });

    it('not value dep while fileName in store', () => {
      const field = [
        {
          name: 'input',
          label: 'input',
          field: 'input',
        },
        {
          name: 'select',
          label: 'select',
          field: 'select',
          dependency: {
            value: {
              relates: ['input'],
              inputs: [['aaa']],
              output: 'one',
              defaultOutput: '',
            },
          },
        },
      ];

      store.parse(field[1], field[1].dependency, getName);
      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'aaa'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'aaa'),
        getInternalHooks: () => {
          return depForm;
        },
      };
      const selectType = store.depsByFieldNameList.get(['input']);
      store.triggerDependency({ form }, selectType, {
        cascadeTrigger: 'setFieldsValue',
        cascadeStore: { select: 'a' },
      });
      expect(form.setFieldValue).not.toHaveBeenCalled();
    });

    it('source function type', () => {
      const field = {
        name: 'select',
        label: 'select',
        field: 'select',
        dependency: {
          source: {
            relates: ['input'],
            autoResetValue: true,
            type: () => {},
          },
        },
      };

      store.parse(field, field.dependency, getName);
      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'aaa'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'aaa'),
        getInternalHooks: () => {
          return depForm;
        },
      };
      const selectType = store.depsByFieldNameList.get(['input']);
      store.triggerDependency({ form }, selectType, {});
      expect(depForm.setFieldValueByFieldKey).toHaveBeenCalledWith(['select'], undefined);
    });

    it('remoteSource', async () => {
      const field = {
        name: 'select',
        label: 'select',
        field: 'select',
        remoteSource: {
          url: '/source.json',
          method: 'post',
        },
        dependency: {
          source: {
            relates: ['input'],
            defaultOutput: [],
          },
        },
      };

      store.parse(field, field.dependency, getName);
      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'input'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'input'),
        getInternalHooks: () => {
          return depForm;
        },
      };
      await store.triggerDependency({ form }, store.depsByFieldNameList.get(['input']), {});

      return store
        .cascadeSource(
          { form },
          {
            name: ['select'],
            relates: [['input']],
            autoResetValue: true,
            remoteSource: {
              url: '/source.json',
              method: 'post',
            },
          },
          {},
        )
        .then(() => {
          expect(form.setFieldValue).toHaveBeenCalledWith(['select'], undefined);
          expect(form.setFieldSource).toHaveBeenCalledWith(['select'], [{ text: 'a', value: 'a' }]);
        });
    });

    it('not remoteSource dep', () => {
      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'input'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'input'),
        getInternalHooks: () => {
          return depForm;
        },
      };

      const res = store.cascadeSource(
        { form },
        {
          name: ['select'],
          type: () => false,
          relates: [['input']],
          autoResetValue: true,
          remoteSource: {
            url: '/source.json',
            method: 'post',
          },
        },
        {},
      );
      expect(form.setFieldSource).not.toHaveBeenCalled();
      expect(res).toBeUndefined();
    });

    it('ignore inputs', () => {
      const field = {
        name: 'select',
        label: 'select',
        field: 'select',
        dependency: {
          source: {
            relates: ['input'],
            autoResetValue: false,
            inputs: ['aaa'],
            ignores: [['a']],
            defaultOutput: [],
            output: [{ text: 'a', value: 'a' }],
          },
        },
      };

      store.parse(field, field.dependency, getName);
      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'a'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'a'),
        getFieldVisible: jest.fn(),
        getInternalHooks: () => {
          return depForm;
        },
      };
      const selectType = store.depsByFieldNameList.get(['input']);
      store.triggerDependency({ form }, selectType, {});
      expect(form.setFieldValue).not.toHaveBeenCalled();
      expect(form.setFieldSource).toHaveBeenCalledWith(['select'], []);
    });

    it('cases', () => {
      const fields = [
        {
          name: 'input',
          label: 'input',
          field: 'input',
        },
        {
          name: 'input2',
          label: 'input2',
          field: 'input',
          dependency: {
            value: {
              relates: ['input'],
              cases: [
                {
                  inputs: [['aaa']],
                  output: 'one',
                },
                {
                  inputs: [['bbb']],
                  output: 'two',
                },
              ],
              defaultOutput: '',
            },
          },
        },
        {
          name: 'input3',
          label: 'input3',
          field: 'input',
          dependency: {
            value: {
              relates: ['input'],
              cases: [
                {
                  inputs: [['aaa']],
                  output: 'one',
                },
                {
                  inputs: [['bbb']],
                  output: 'two',
                },
              ],
              defaultOutput: '',
            },
          },
        },
        {
          name: 'input4',
          label: 'input4',
          field: 'input4',
          dependency: {
            value: {
              relates: ['input'],
              inputs: [['aaa']],
              output: '4',
              defaultOutput: '',
            },
            visible: {
              relates: ['input'],
              inputs: [['aaa']],
              output: true,
              defaultOutput: false,
            },
          },
        },
      ];

      store.parse(fields[1], fields[1].dependency, getName);
      store.parse(fields[2], fields[2].dependency, getName);
      store.parse(fields[3], fields[3].dependency, getName);

      const inputType = store.depsByFieldNameList.get(['input']);

      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'aaa'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'aaa'),
        getFieldVisible: jest.fn(),
        getInternalHooks: () => {
          return depForm;
        },
      };
      store.triggerDependency({ form }, inputType, {});
      expect(depForm.setFieldValueByFieldKey.mock.calls).toEqual([
        [['input2'], 'one'],
        [['input3'], 'one'],
        [['input4'], '4'],
      ]);
      expect(form.getFieldVisible).toHaveBeenCalledWith(['input4']);
      expect(form.setFieldVisible).toHaveBeenCalledWith(['input4'], true);
    });
  });

  describe('utils', () => {
    it('hasOwnPropWithNameList', () => {
      expect(hasOwnPropWithNameList(['input'], { input: 'aaa' })).toBeTruthy();
      expect(hasOwnPropWithNameList(['input'], { input2: 'aaa' })).toBeFalsy();
    });

    it('equal', () => {
      expect(equal('*', [1, 2, 3])).toBeTruthy();
      expect(equal([1, 2, 3], 1)).toBeTruthy();
      expect(equal([1, 2, 3], 0)).toBeFalsy();
    });

    it('single match', () => {
      expect(
        singleMatch(
          [
            [1, 2],
            [2, 3],
          ],
          [1, 4],
        ),
      ).toBeTruthy();
      expect(
        singleMatch(
          [
            [1, 2],
            [2, 3],
          ],
          [4],
        ),
      ).toBeFalsy();
    });

    it('all match', () => {
      expect(
        allMatch(
          [
            [1, 2],
            [2, 3],
          ],
          [2, 2],
        ),
      ).toBeTruthy();
      expect(
        allMatch(
          [
            [1, 2],
            [2, 3],
          ],
          [1, 4],
        ),
      ).toBeFalsy();
    });
  });

  describe('FormDependency', () => {
    let formDependency;
    beforeEach(() => {
      formDependency = new FormDependency();
    });

    it('parseFormDependency', () => {
      expect(
        formDependency.parseFormDependency(
          {
            name: 'input2',
            label: 'input2',
            field: 'input',
            dependency: {
              value: {
                relates: ['input'],
                inputs: [['aaa']],
                output: 'one',
                defaultOutput: '',
              },
            },
          },
          getName,
        ),
      ).toEqual({ name: 'input2', label: 'input2', field: 'input' });
      expect(formDependency.depStore.depsByFieldNameList.getNameLists()).toEqual([['input']]);
      expect(formDependency.getCascades()).toEqual([['input']]);
      const depForm = {
        getFieldValueByFieldKey: jest.fn(() => 'aaa'),
        setFieldValueByFieldKey: jest.fn(),
        getFieldKeyByFieldName: jest.fn((a) => a),
        getFieldNameByFieldKey: jest.fn((a) => a),
        getPrevFieldNameByFieldKey: jest.fn((a) => a),
      };

      const form = {
        setFieldDisabled: jest.fn(),
        setFieldVisible: jest.fn(),
        setFieldValue: jest.fn(),
        setFieldSource: jest.fn(),
        getFieldValue: jest.fn(() => 'aaa'),
        getInternalHooks: () => {
          return depForm;
        },
      };
      formDependency.cascade({ form }, [{ name: ['select'], value: 'a' }], {});
      expect(form.setFieldValue).not.toHaveBeenCalled();
      form.setFieldValue.mockRestore();

      formDependency.cascade({ form }, [{ name: ['input'], value: 'aaa' }], {});
      expect(depForm.setFieldValueByFieldKey).toHaveBeenCalledWith(['input2'], 'one');
      form.setFieldValue.mockRestore();
    });

    it('empty dependency', () => {
      expect(
        formDependency.parseFormDependency(
          {
            name: 'input2',
            label: 'input2',
            field: 'input',
            dependency: {},
          },
          getName,
        ),
      ).toEqual({ name: 'input2', label: 'input2', field: 'input', dependency: {} });
      expect(formDependency.depStore.depsByFieldNameList.getNameLists()).toEqual([]);
    });
  });
});
