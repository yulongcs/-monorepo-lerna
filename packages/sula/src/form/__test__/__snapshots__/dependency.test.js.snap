// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`depstore store basic 1`] = `
Object {
  "disabled": Array [
    Array [
      Object {
        "autoResetValue": true,
        "defaultOutput": false,
        "ignores": Array [
          Array [
            undefined,
          ],
        ],
        "inputs": Array [
          Array [
            "disabled",
          ],
        ],
        "name": Array [
          "input2",
        ],
        "output": true,
        "relates": Array [
          Array [
            "input",
          ],
        ],
        "type": undefined,
      },
    ],
  ],
  "source": Array [
    Array [
      Object {
        "autoResetValue": true,
        "defaultOutput": Array [],
        "ignores": Array [
          Array [
            undefined,
          ],
        ],
        "inputs": Array [
          Array [
            "source",
          ],
        ],
        "name": Array [
          "input2",
        ],
        "output": Array [],
        "relates": Array [
          Array [
            "input",
          ],
        ],
        "remoteSource": Object {
          "method": "post",
          "url": "/source.json",
        },
        "type": undefined,
      },
    ],
  ],
  "value": Array [
    Array [
      Object {
        "autoResetValue": true,
        "defaultOutput": "",
        "ignores": Array [
          Array [
            undefined,
          ],
        ],
        "inputs": Array [
          Array [
            "aaa",
          ],
        ],
        "name": Array [
          "input2",
        ],
        "output": "one",
        "relates": Array [
          Array [
            "input",
          ],
        ],
        "type": undefined,
      },
    ],
  ],
  "visible": Array [
    Array [
      Object {
        "autoResetValue": true,
        "defaultOutput": true,
        "ignores": Array [
          Array [
            undefined,
          ],
        ],
        "inputs": Array [
          Array [
            "hidden",
          ],
        ],
        "name": Array [
          "input2",
        ],
        "output": false,
        "relates": Array [
          Array [
            "input",
          ],
        ],
        "type": undefined,
      },
    ],
  ],
}
`;
